shell/platform/common/cpp/text_input_model.cc

```diff
@@ -76,7 +84,7 @@ void TextInputModel::DeleteSelected() {
   selection_extent_ = selection_base_;
 }
 
-void TextInputModel::AddCharacter(char c) {
+void TextInputModel::AddCharacter(uint32_t c) {
```

Comment:
Why uint32_t rather than char32_t? Same question for the other APIs.

Reply:
char32_t is the same as uint_least32_t, while uint32_t is absolutely 32-bit. In the API interface, I think it's clearer to use uint32_t.

Reply:
> char32_t is the same as uint_least32_t, while uint32_t is absolutely 32-bit.

So on a theoretical system with no 32-bit integers, you want this code to fail to compile rather than use the best available type? I don't see how that's desirable.

> In the API interface, I think it's clearer to use uint32_t.

Why? What's important at the API level is that it's a UTF-32 character, not that it's exactly 32 bits. `char32_t` is exactly the type that expresses conceptually what this interface accepts.

Reply:
> char32_t
In my opinion, until “AddCharacter” is called, we can't be sure that it must be a char32_t.
It's probably a UTF-16 (Windows + WM_CHAR) or a UTF-32 (Windows + WM_UNICHAR). As I mentioned in response to your question about how to deal with WM_CHAR, I think this logic is better in “AddCharacter”. If this logic is processed in win32/glfw, the API should be changed to char32_t.

Reply:
> It's probably a UTF-16 (Windows + WM_CHAR) or a UTF-32 (Windows + WM_UNICHAR).

We know exactly which it is at the point where we receive the message; we should do any necessary conversion at that point, rather than discard the information and then try to re-infer it later.

> If this logic is processed in win32/glfw

I'm confused by the reference to GLFW here and in your comment. Why would the GLFW code need UTF-16 handling?

Reply:
third_party\glfw\src\win32_window.c
flutter\shell\platform\windows\win32_window.cc

If we put the logic on the first site, we need to modify it in the message loop(WindowProc) in both files.

Which one is our required? Currently, it only run in glfw/src/win32_windows.cc when I debug FDE.

Reply:
> Which one is our required? Currently, it only run in glfw/src/win32_windows.cc when I debug FDE.

You need to update your FDE and Flutter checkouts then; we recently switched from GLFW to the Win32 embedding. At this point we are no longer actively supporting the GLFW embedding on Windows.

> Also, glfw is third-party code, and we shouldn't change it too much.

We shouldn't change it at all. If we were still using GLFW on Windows and the GLFW Windows code isn't doing what the documentation says, then we would want to fix it upstream.

Reply:
DONE.